packages/core-utils/src/browser-storage.ts

```diff
@@ -0,0 +1,174 @@
+        return;
+      }
+      const [chosenKey] = _.head(
+        _.sortBy(([_k, v]: [any, any]) => {
+          const data: any = maybeParseJSON(v);
+          return data && _.isInteger(data.timestamp) ? data.timestamp : -Infinity;
+        }, candidates)
+      ) as [any];
+      storage.removeItem(chosenKey);
+    }
+  }
+};
+
+export const getStatic = (storage: any, key: any) => {
+  return maybeParseJSON(storage.getItem(key));
+};
+
+/**
+ * Sets a key/value in browser storage, or removes the key from storage if `value` is undefined.
+ *
+ * Static storage will not be automatically deleted in the case of space overflow. If your storage usage
+ * is transient (for example, navigating between routes and preserving data where the loss of it is not
+ * critical), instead use `setDynamic`.
+ *
+ * See also `staticStorageSlot`, which can be passed to the `useStore` hook function.
+ *
+ * @param storage an object implementing the Storage interface
+ * @param key the key for storing the value
+ * @param value the new JSON-serializable value to be stored, or `undefined` to remove the key.
+ */
+export const setStatic = (storage: any, key: any, value: any) => {
+  if (value === undefined) {
+    storage.removeItem(key);
+  } else {
+    forceSetItem(storage, key, JSON.stringify(value));
+  }
+};
+
+export const listenStatic = (storage: any, key: any, fn: any) => {
+  window.addEventListener('storage', (e) => {
+    if (e.storageArea === storage && e.key === key && e.newValue) {
+      fn(maybeParseJSON(e.newValue));
+    } else {
+      fn(undefined);
+    }
+  });
+};
+
+export const getDynamic = (storage: any, key: any) => {
+  const storageKey = `dynamic-storage/${key}`;
+  const data: any = maybeParseJSON(storage.getItem(storageKey));
+  return data?.value;
+};
+
+/**
+ * Sets a key/value in browser storage, or removes the key from storage if `value` is undefined.
+ *
+ * Dynamic storage will be automatically deleted in the case of space overflow, deleting oldest stored
+ * data first. If your storage usage is not transient, instead use `setStatic`.
+ *
+ * @param storage an object implementing the Storage interface
+ * @param key the key for storing the value. Note that it will be prefixed by a value to allow separating
+ * static from dynamic storage.
+ * @param value the new JSON-serializable value to be stored, or `undefined` to remove the key
+ */
+export const setDynamic = (storage: any, key: any, value: any) => {
+  const storageKey = `dynamic-storage/${key}`;
+  if (value === undefined) {
+    storage.removeItem(storageKey);
+  } else {
+    forceSetItem(storage, storageKey, JSON.stringify({ timestamp: Date.now(), value }));
+  }
+};
+
+export const listenDynamic = (storage: any, key: any, fn: any) => {
+  const storageKey = `dynamic-storage/${key}`;
+  window.addEventListener('storage', (e) => {
+    if (e.storageArea === storage && e.key === storageKey && e.newValue) {
```

Comment:
Added check for e.newValue, because maybeParseJSON expects a string (not null or undefined).

However, since this was being called in Javascript before (without type checks), I am using undefined in the else block, because that's what maybeParseJSON's catch block does.

Reply:
see event type comment above... same here.